lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3157,20 +3157,43 @@ DynamicallyReplacedDeclRequest::evaluate(Evaluator &evaluator,
   return nullptr;
 }
 
-/// Returns true if the given type conforms to `Differentiable` in the given
-/// module.
-static bool conformsToDifferentiable(Type type, DeclContext *DC) {
+/// If the given type conforms to `Differentiable` in the given context, returns
+/// the `ProtocolConformanceRef`. Otherwise, returns an invalid
+/// `ProtocolConformanceRef`.
+///
+/// This helper verifies that the `TangentVector` type witness is valid, in case
+/// the conformance has not been fully checked and the type witness cannot be
+/// resolved.
+static ProtocolConformanceRef getDifferentiableConformance(Type type,
+                                                           DeclContext *DC) {
   auto &ctx = type->getASTContext();
   auto *differentiableProto =
       ctx.getProtocol(KnownProtocolKind::Differentiable);
-  auto conf = TypeChecker::conformsToProtocol(
-      type, differentiableProto, DC, ConformanceCheckFlags::InExpression);
+  auto conf =
+      TypeChecker::conformsToProtocol(type, differentiableProto, DC, None);
   if (!conf)
-    return false;
+    return ProtocolConformanceRef();
   // Try to get the `TangentVector` type witness, in case the conformance has
-  // not been fully checked and the type witness cannot be resolved.
+  // not been fully checked.
+  Type tanType = conf.getTypeWitnessByName(type, ctx.Id_TangentVector);
+  if (tanType.isNull() || tanType->hasError())
```

Comment:
This mix of semantic and syntactico-semantic checks doesn't feel right. Is the idea that you're trying to check for a conformance to this protocol, or to a well-formed conformance of this protocol?

Reply:
We want to check for well-formed conformances to the `Differentiable` protocol, for the purpose of getting a valid `TangentVector` associated type witness. If we don't verify that type witnesses are valid, the compiler crashes downstream: https://github.com/apple/swift/pull/30936#discussion_r406537011

Do you have any change suggestions?

Reply:
It feels like `DifferentiableAttributeTypeCheckRequest` (which needs to be broken down) needs to verify this as a precondition before it calls through to the rest of these functions, which need to assert this as a precondition. Because you don't ever *need* the witness, just its well-formedness.